Pod/Classes/BNCBranchIntegration.m

```diff
@@ -1,61 +1,310 @@
+        BNCProductCategoryLuggageBags,
+        BNCProductCategoryMature,
+        BNCProductCategoryMedia,
+        BNCProductCategoryOfficeSupplies,
+        BNCProductCategoryReligious,
+        BNCProductCategorySoftware,
+        BNCProductCategorySportingGoods,
+        BNCProductCategoryToysGames,
+        BNCProductCategoryVehiclesParts,
+    ];
+    
+    for (BNCProductCategory category in categories) {
+        if ([string compare:category options:NSCaseInsensitiveSearch|NSDiacriticInsensitiveSearch] == NSOrderedSame)
+            return category;
+    }
+    return /*nil*/ BNCProductCategoryBabyToddler;   // eDebug - test this.
+}
+
+#define addStringField(field, name) { \
+    NSString *value = dictionary[@#name]; \
+    if (value) { \
+        if ([value isKindOfClass:NSString.class]) \
+            field = value; \
+        else \
+            field = [value description]; \
+        dictionary[@#name] = nil; \
+    } \
+}
+
+#define addDecimalField(field, name) { \
+    NSString *value = dictionary[@#name]; \
+    if (value) { \
+        if (![value isKindOfClass:NSString.class]) \
+            value = [value description]; \
+        field = [NSDecimalNumber decimalNumberWithString:value]; \
+        dictionary[@#name] = nil; \
+    } \
+}
+
+#define addDoubleField(field, name) { \
+    NSNumber *value = dictionary[@#name]; \
+    if ([value respondsToSelector:@selector(doubleValue)]) { \
+        field = [value doubleValue]; \
+        dictionary[@#name] = nil; \
+    } \
+}
+
++ (BranchUniversalObject*) universalObjectFromDictionary:(NSMutableDictionary*)dictionary {
+    NSInteger initialCount = dictionary.count;
+    if (initialCount == 0) return nil;
+
+    /* Segment product fields:
+        product_id
+        sku
+        category
+        name
+        brand
+        variant
+        price
+        quantity
+        url
+        image_url
+    */
+
+    BranchUniversalObject *object = [[BranchUniversalObject alloc] init];
+    object.contentMetadata.contentSchema = BranchContentSchemaCommerceProduct;
+    
+    addStringField(object.canonicalIdentifier, product_id);
+    addStringField(object.contentMetadata.sku, sku);
+    BNCProductCategory category = [self categoryFromString:dictionary[@"category"]];
+    if (category) {
+        object.contentMetadata.productCategory = category;
+        dictionary[@"category"] = nil;
+    }
+    addStringField(object.contentMetadata.productName, name);
+    addStringField(object.contentMetadata.productBrand, brand);
+    addStringField(object.contentMetadata.productVariant, variant);
+    addDecimalField(object.contentMetadata.price, price);
+    addDoubleField(object.contentMetadata.quantity, quantity);
+    addStringField(object.canonicalUrl, url);
+    addStringField(object.imageUrl,  image_url);
+
+    // Segment fields not handled: coupon and position, for instance.
+    object.contentMetadata.customMetadata = [self stringDictionaryFromDictionary:dictionary];
```

Comment:
Will this add the entire dictionary to the custom metadata or only the fields that cannot be mapped to Branch  defined keys ?

Reply:
It only adds the fields that can't be mapped to Branch defined keys.